google/cloud/aiplatform/datasets/datasources.py

```diff
@@ -0,0 +1,140 @@
+import abc
+from typing import Optional, Dict, Sequence
+from google.cloud.aiplatform_v1beta1 import (
+    GcsSource,
+    ImportDataConfig,
+)
+from google.cloud.aiplatform import utils
+
+
+class _Datasource(abc.ABC):
+    """An abstract class that sets dataset_metadata"""
```

Comment:
How much of the distinction between tabular vs non-tabular, and importable vs non-importable is tentative oddity that will be removed from the service, and as such how much should be completely hidden from the user?

As an example, while tabular datasets do not support an import_data service method, the library could well support it through BQ/GCS, which is the expected mechanism for updating tabular datasets.

Reply:
The near-term goal is to keep the Tabular vs Non-tabular distinction internal so the users don't have to worry about it if they used the Dataset subclasses. The subclasses will tell them exactly which arguments are relevant.

However, there are always going to be some distinctions between tabular vs non-tabular such as bq_source_uris won't likely be supported by non-tabular datasources.

In the case of importing data, once we have a mechanism to import data to tabular datasets, we just conform TabularDatasource to  _DatasourceImportable, provide the relevant logic and it should thus gain the ability to import data.